File: Lab1_WFApp/Lab1_WFApp/Classes/Rule.cs
using System.Collections.Generic;
using System.Linq;

namespace Lab1_WFApp
{
    class Rule
    {
        //условие
        public List<string> Antecedent { get; set; }
        //текст
        public string Ask { get; set; }
        //действие
        public string Consequent { get; set; }

        public Rule(List<string> Antecedent, string Ask, string Consequent)
        {
            this.Antecedent = Antecedent;
            this.Ask = Ask;
            this.Consequent = Consequent;
        }
        public Rule(Rule rule)
        {
            this.Antecedent = rule.Antecedent;
            this.Ask = rule.Ask;
            this.Consequent = rule.Consequent;
        }
        public Rule()
        {
        }


        public float getSimilarDegree(Rule rule)
        {
            float degree = 0;

            //получаем количество совпадений и вычисляем коэффициент
            List<string> similars = new List<string>(rule.Antecedent);
            //similars.retainAll(Antecedent);
            var sim = from t in rule.Antecedent
                      where Antecedent.Contains(t)
                      select t;
            degree = (float)similars.Count / (float)Antecedent.Count;

            return degree;
        }
       
        public bool isContained(List<Rule> rules)
        {
            bool flag = false;

            foreach (Rule r in rules)
            {
                if (getSimilarDegree(r) == 1 &&
                   r.Ask.Equals(Ask) &&
                   r.Consequent.Equals(Consequent))
                {
                    flag = true;
                    break;
                }
            }
            return flag;
        }
    }
}

File: Lab1_java_v3/src/Classes/Rule.java
package Classes;

import java.util.ArrayList;
import java.util.List;
/**
 * Класс "Правило": условия, текст, результат (не книга)
 */
public class Rule {

    //условие
    private List<String> antecedent;
    //текст
    private String ask;
    //действие
    private String consequent;

    public Rule(List<String> antecedent, String ask, String consequent) {
        this.antecedent = antecedent;
        this.ask = ask;
        this.consequent = consequent;
    }
    public Rule(Rule rule) {
        this.antecedent = rule.antecedent;
        this.ask = rule.ask;
        this.consequent = rule.consequent;
    }
    public Rule() {
    }

    public List<String> getAntecedent() {
        return antecedent;
    }

    public String getAsk() {
        return ask;
    }

    public String getConsequent() {
        return consequent;
    }


    public float getSimilarDegree(Rule rule){
        float degree = 0;

        //получаем количество совпадений и вычисляем коэффициент
        List<String> similars = new ArrayList<>(rule.getAntecedent());
        similars.retainAll(antecedent);
        degree = (float)similars.size() / (float)antecedent.size();

        return  degree;
    }

    public boolean isContained(List<Rule> rules){
        boolean flag = false;

        for (Rule r: rules) {
            if(getSimilarDegree(r) == 1 &&
               r.getAsk().equals(ask)&&
               r.getConsequent().equals(consequent)){
                flag = true;
                break;
            }
        }
        return flag;
    }
}
